### PR TITLE
Pipeline support for changeset pull and configurable workspace cleanup

### PR DESCRIPTION
**Pull request by knapsu**
Fixes and improvements from [Housemarque](https://housemarque.com/).

- `Improvement` _Pipeline_ jobs can now pull from a changeset number.
- `Improvement` Make it possible to define cleanup method for each workspace when doing an update.
- `Improvement` Prevent cluttering console output when retrieving changesets history.
- `Fix` Workspace directories can now be nested multiple levels deep.
- Other small improvements and fixes.

### CODE REVIEW

**Comment by knapsu** [comment]
update_cleanup_01

**Comment by mig42** [comment]
Hi @knapsu!

I'm sorry, I missed this one 😯 I'll try to review and test it as soon as I can! Thank you for the effort 😊

**Comment by knapsu** [comment]
@mig42 I've merged master branch. Please review.

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java:248]: I'd prefer not to change the object state in this method - it's kind of obscure to me. I'd rather have this code in the calling method (`checkout()`).

If you really want to have this isolated, you could rename the method to something like `updateCleanupMethodFromOlderConfig()` or similar so one doesn't need to go inside the method to find out what it does.

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticTool.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/PlasticTool.java:101]: Following PR #36 and #37 , the changeset log command doesn't bloat the build output anymore. Does this change make sense now? Since all XML-based commands (find & log) output to a file, I'm not sure if we'll need this...

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:34]: I noticed that you changed some method signatures that have one parameter per line to have all parameters in the same line.

I think it's a better idea to avoid changing the style in this big PRs because it can add noise. We could discuss creating a PR specifically to unify style and so on, but I believe this style-only changes should be avoided if possible.

Besides, this line (and some others) exceed the 120-character limit!

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:52]: This comment would be better placed before the `.setSelector()` call!

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:53]: We can get rid of this line 😛

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:88]: Question: Does this really need to be protected?

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:75]: I think a better name for this would be `parentWorkspaces`, right?

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:143]: I think a better name for this would be `findParentWorkspaces()`, right?

File: src/main/java/com/codicesoftware/plugins/hudson/commands/GetSelectorSpecCommand.java
Comment (mig42) [inline src/main/java/com/codicesoftware/plugins/hudson/commands/GetSelectorSpecCommand.java:22]: This got my attention: in other instances you change abbreviated commands to use the full name, but this case is the other way around. Shouldn't we unify that?

File: src/main/webapp/cleanup.html
Comment (mig42) [inline src/main/webapp/cleanup.html:3]: I'd prefer this long HTML lines to be cut off so they don't exceed the 120-char limit. We prefer to be able to read the code without horizontal scroll or line wrapping 😊

File: src/test/java/com/codicesoftware/plugins/hudson/PlasticSCMTest.java
Comment (mig42) [inline src/test/java/com/codicesoftware/plugins/hudson/PlasticSCMTest.java:19]: We should remove the test completely rather than removing the `@Test` annotation, right? Otherwise I prefer to have it marked as test but ignored, as it was before.

**Review by mig42 — CHANGES_REQUESTED** [review]
Thanks for the contribution @knapsu ! I'd like to perform some small changes before merging this PR to main.

I also have a question regarding this:
> `Fix` Workspace directories can now be nested multiple levels deep

What's the use case for that? The `Directory` setting in the config doesn't allow to specify subdirectories, so I'm not exactly sure about what this is fixing...

About this:
> `Improvement` Prevent cluttering console output when retrieving changesets history

This issue went away with PRs #36  and #37, so it might be a good idea to either remove it from the changes list or even remove it completely.

Finally, as a side note, I think it would be a good idea in the future to submit smaller, multiple PRs (each one with a single improvement/fix) instead of a big one. I believe that we could all benefit from that 😊 That way it's easier for us to allocate time to review and validate it. This increases the chances that we process PRs earlier and have them merged. It decouples fixes too, so change requests during review don't put on hold a bunch of new features, just one of them. What do you think?

**Comment by knapsu** [comment]
Hi.

> I also have a question regarding this:
> 
> > `Fix` Workspace directories can now be nested multiple levels deep
> 
> What's the use case for that? The `Directory` setting in the config doesn't allow to specify subdirectories, so I'm not exactly sure about what this is fixing...

This corrects a bug where the plugin would fail if there would be an existing workspace located in for example `WORKSPACE/libraries/SDL`). You correctly said that using "Directory" field would not allow such thing but there are also custom shell scripts which pull third party libraries to other locations or just to do partial checkout which plugin does not suppor. With this PR it will discover such workspaces and handle those gracefully.

> About this:
> 
> > `Improvement` Prevent cluttering console output when retrieving changesets history
> 
> This issue went away with PRs #36 and #37, so it might be a good idea to either remove it from the changes list or even remove it completely.
> 

This is a different thing. #36 and it's fix in #37 is for command memory usage (which was filled with the big chunks of XML text output). My changes add a possibility for commands to be quiet and skip outputting the Plastic client communication to the console to avoid clutter. For example now retrieving log history between two changesets will be silenced. No one really wants to see XML in the output. They want to see a pretty and neat list of changes in the job summary instead (sometimes just the XML had a couple of megabytes of text - some drastic cases had even 30 MB!).

> Finally, as a side note, I think it would be a good idea in the future to submit smaller, multiple PRs (each one with a single improvement/fix) instead of a big one. I believe that we could all benefit from that blush That way it's easier for us to allocate time to review and validate it. This increases the chances that we process PRs earlier and have them merged. It decouples fixes too, so change requests during review don't put on hold a bunch of new features, just one of them. What do you think?

Yup, smaller PRs would be good, but the reality is that when I start working on one I see something broken and just start patching/improving it immediately. This is how from a single thing it becomes more and more. Doing tests is time consuming as I do test not only locally but also on real data/use cases. That is why I prefer to check multiple things with one iteration, because of the time spend on settign up the tests. Also I had few cases where with multiple PR's one would be dependent on fixes provided in earlier PR. And it is impossible to create a PR in GitHub in this situation where necessary changes are not yet in (or is it ?). Even if it would be possible, merging each change in master or other PR's back to every PR that is waiting would be a real burden for me. Either way I will do my best to make PR's smaller to make them easier to digest. :+1:

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java:705]: To be honest I do not remember now. IT was so long time ago. Probably just to make the code look similar to the rest of the classes. The `final` here does not seem to do any real importance here. Should I restore it?

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java:248]: I do not understand what you mean. This is the `checkout(...)` method which is the entry point for Jenkins SCM integrations. Where would you want it to be moved?

Regarding the name, my idea was to have a method that checks if job config is in obsolete format and applies it's settings to currently existing formula. I am aware that there is now just one field, the cleanup method, but the name was made with planning ahead, that there might be more fields like this.

Maybe leave this method and rename it to something more explanatory and do more verbose javadoc comment? `adjustConfigurationFromOlderFormat(...)` ?

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticTool.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/PlasticTool.java:101]: Like mentioned in the top comment, this is something different. PR #36 and it's fix in #37 seems to be for improving memory usage. The Jenkins console output will still be cluttered with XML text. This improvement allows to execute commands in quiet mode.

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:34]: The final code style based on official Sun and Google conventions was planned to be in next PR which I wanted to discuss together with you after this was merged. Will revert this. Checkstyle schema in next PR will enforce whichever form we choose.

Please not we are using **160** character limit. You are saying about 120 characters but if you would run checkstyle on Maven verify phase you would see there is no problems with the code.

We are in XXI century. Let's not get medieval about character limits. 80 character enforcement are in my opinion a joke. 120 is good but with everyone now having Full HD monitors or even bigger ones we should not get crazy. Too narrow code is not friendly. Also huge on liners are not. Something to discuss in next PR.

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:75]: Parent would indicate it is the immediate directory above (1 level up). This list will however gather all workspaces which are above this one (all levels up).

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:143]: It's a matter of what is more understandable and have less ambiguity. Inner/Outer or Child/Parent. I used inner/outer and child/parent in the comments to ensure it would be understandable what this fragment does despite the disambiguity of the child/parent name. Would leave it but if you feel differently then maybe change it?

```
        // Handle situation where workspace exists in child path.
        List<Workspace> innerWorkspaces = findWorkspacesInsidePath(workspacePath, workspaces);
        for (Workspace workspace : innerWorkspaces) {
            deleteWorkspace(tool, workspace, workspaces);
        }

        // Handle situation where workspace exists in parent path.
        List<Workspace> outerWorkspaces = findWorkspacesOutsidePath(workspacePath, workspaces);
        for (Workspace workspace : outerWorkspaces) {
            deleteWorkspace(tool, workspace, workspaces);
        }
```

File: src/main/java/com/codicesoftware/plugins/hudson/commands/GetSelectorSpecCommand.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/commands/GetSelectorSpecCommand.java:22]: Yes. Correcting this.

File: src/main/webapp/cleanup.html
Comment (knapsu) [reply src/main/webapp/cleanup.html:3]: Here is something that I do not fully agree. It's XXI century and IDE's and text editors has the functionality of wrapping long lines. But still I understand your point. Changing this.

File: src/test/java/com/codicesoftware/plugins/hudson/PlasticSCMTest.java
Comment (knapsu) [reply src/test/java/com/codicesoftware/plugins/hudson/PlasticSCMTest.java:19]: If the test is ignored then running Maven test phase will report this test as skipped, giving a warning. This looks bad. The test was ignored because depending on the test environment it will have Plastic CLI `cm` installed or not. We are unsure of this and this test can not be trusted. I kept this test for my own checks. Will remove it as it is not generic enough to be properly useful.

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java]: I noticed that you rolled that back. Thanks!

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/PlasticSCM.java]: I think the new name is better indeed. Maybe we can improve this in the future, but it's good for the time being 😊

File: src/main/java/com/codicesoftware/plugins/hudson/PlasticTool.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/PlasticTool.java:101]: Hi,

It's not accurate to say that the Jenkins console output will still be cluttered with XML text. Commands `cm log` or `cm find` won't print anything to standard output if you tell them to write the XML output to a file.

We'll accept these changes and we'll evaluate their performance in the future to study their redundancy.

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java]: Thanks for rolling those back.

My request was not about being medieval -as you said-, but to make it easier for anyone to review code. For instance, when I open the GitHub website, line width is about 110 characters. We should avoid making people scroll horizontally to read/review PRs.

I agree that it will be a good idea to decide a common style in a future PR and get it enforced during the maven build process.

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:87]: We'll accept this change. We'll think if we can come up with a better naming 😉

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:155]: We'll accept this name. We'll think if we can come up with a better naming 😉

Also, we're worried about the change from 'private' to 'protected' in this static class. We'll evaluate it and probably submit a new PR.

File: src/main/java/com/codicesoftware/plugins/hudson/commands/GetSelectorSpecCommand.java
Comment (mig42) [reply src/main/java/com/codicesoftware/plugins/hudson/commands/GetSelectorSpecCommand.java]: Ok, thanks for the update. We'll probably come up with some guidelines for this kind of stuff in the near future - e.g. always use full name commands instead of abbreviations.

File: src/main/webapp/cleanup.html
Comment (mig42) [reply src/main/webapp/cleanup.html]: Cool, thanks! As you said, this is not about working in an IDE but rather about displaying code in the GitHub website or diffing changes.

**Review by mig42 — APPROVED** [review]
We'll go ahead with this PR! We have some concerns about some bits of it, but we'll study how to improve those and come up with a proposal via PR.

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:100]: Yes. Please see  `CheckoutActionTest.java`. If it is private then I will not be able to access this method and do unit tests. The same goes for other methods that are under tests for this class.

It could be also without modifier (the package scope) however I prefer to always explicitly specify the accessibility scope as it makes it easier to notice that the method/field is accessed somewhere else.

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java:155]: So maybe `allChildWorkspaces` and `allParentWorkspaces` will be better?

File: src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java
Comment (knapsu) [reply src/main/java/com/codicesoftware/plugins/hudson/actions/CheckoutAction.java]: Yes, I apologize and totally agree with you. Later after writing my comment I started to think of situations where actually IDE is not used and line wrapping might be missing. Unfortunately there still such cases exist (mainly text diff tools) so keeping it 120 is the best choice. GitHub web is also a good example and is kind of unexpected as HTML text can be nicely wrapped in the browsers.

What I meant was that some tools are still in the past century, not us :) But you are right, we should support such cases. :+1:

**Review by alecharp — APPROVED** [review]

**Pull request merged**
